File: src/main/java/com/saomc/saoui/GLCore.java
package com.saomc.saoui;

import com.saomc.saoui.util.ColorUtil;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.FontRenderer;
import net.minecraft.client.renderer.GlStateManager;
import net.minecraft.client.renderer.Tessellator;
import net.minecraft.client.renderer.texture.TextureManager;
import net.minecraft.client.renderer.vertex.DefaultVertexFormats;
import net.minecraft.client.renderer.vertex.VertexFormat;
import net.minecraft.client.resources.IReloadableResourceManager;
import net.minecraft.client.settings.GameSettings;
import net.minecraft.util.ResourceLocation;
import net.minecraft.util.math.AxisAlignedBB;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;
import org.lwjgl.opengl.GL11;

@SideOnly(Side.CLIENT)
public final class GLCore {

    private static Minecraft mc = Minecraft.getMinecraft();

    private GLCore() {
    }

    private static FontRenderer glFont() {
        return mc.fontRenderer;
    }

    private static TextureManager glTextureManager() {
        return mc.getTextureManager();
    }

    public static void glColor(float red, float green, float blue) {
        GlStateManager.color(red, green, blue);
    }

    public static void glColor(float red, float green, float blue, float alpha) {
        GlStateManager.color(red, green, blue, alpha);
    }

    public static void glColorRGBA(ColorUtil color) {
        glColorRGBA(color.rgba);
    }

    public static void glColorRGBA(int rgba) {
        final float red = (float) ((rgba >> 24) & 0xFF) / 0xFF;
        final float green = (float) ((rgba >> 16) & 0xFF) / 0xFF;
        final float blue = (float) ((rgba >> 8) & 0xFF) / 0xFF;
        final float alpha = (float) ((rgba) & 0xFF) / 0xFF;

        glColor(red, green, blue, alpha);
    }

    private static int glFontColor(int rgba) {
        final int alpha = (rgba) & 0xFF;
        final int red = (rgba >> 24) & 0xFF;
        final int blue = (rgba >> 8) & 0xFF;
        final int green = (rgba >> 16) & 0xFF;

        return (alpha << 24) | (red << 16) | (blue << 8) | (green);
    }

    public static void glString(FontRenderer font, String string, int x, int y, int argb, boolean shadow) {
        if (font != null) font.drawString(string, x, y, glFontColor(argb), shadow);
    }

    public static void glString(FontRenderer font, String string, int x, int y, int argb) {
        glString(font, string, x, y, argb, false);
    }

    public static void glString(String string, int x, int y, int argb, boolean shadow) {
        glString(glFont(), string, x, y, argb, shadow);
    }

    public static void glString(String string, int x, int y, int argb) {
        glString(string, x, y, argb, false);
    }

    public static void setFont(Minecraft mc, boolean custom) {
        ResourceLocation fontLocation = custom ? new ResourceLocation(SAOCore.MODID, "textures/ascii.png") : new ResourceLocation("textures/font/ascii.png");
        mc.fontRenderer = new FontRenderer(mc.gameSettings, fontLocation, mc.getTextureManager(), false);
        if (mc.gameSettings.language != null) {
            mc.fontRenderer.setUnicodeFlag(mc.isUnicode());
            mc.fontRenderer.setBidiFlag(mc.getLanguageManager().isCurrentLanguageBidirectional());
        }
        ((IReloadableResourceManager) mc.getResourceManager()).registerReloadListener(mc.fontRenderer);
    }

    private static int glStringWidth(FontRenderer font, String string) {
        return font != null ? font.getStringWidth(string) : 0;
    }

    public static int glStringWidth(String string) {
        return glStringWidth(glFont(), string);
    }

    private static int glStringHeight(FontRenderer font) {
        return font != null ? font.FONT_HEIGHT : 0;
    }

    public static int glStringHeight() {
        return glStringHeight(glFont());
    }

    private static void glBindTexture(TextureManager textureManager, ResourceLocation location) {
        if (textureManager != null) textureManager.bindTexture(location);
    }

    public static void glBindTexture(ResourceLocation location) {
        glBindTexture(glTextureManager(), location);
    }

    public static void glTexturedRect(double x, double y, double z, double width, double height, double srcX, double srcY, double srcWidth, double srcHeight) {
        float f = 0.00390625F;
        float f1 = 0.00390625F;
        Tessellator tessellator = Tessellator.getInstance();
        tessellator.getBuffer().begin(GL11.GL_QUADS, DefaultVertexFormats.POSITION_TEX);
        tessellator.getBuffer().pos(x, (y + height), z).tex((double) ((float) (srcX) * f), (double) ((float) (srcY + srcHeight) * f1)).endVertex();
        tessellator.getBuffer().pos(x + width, y + height, z).tex((double) ((float) (srcX + srcWidth) * f), (double) ((float) (srcY + srcHeight) * f1)).endVertex();
        tessellator.getBuffer().pos(x + width, y, z).tex((double) ((float) (srcX + srcWidth) * f), (double) ((float) (srcY) * f1)).endVertex();
        tessellator.getBuffer().pos(x, y, z).tex((double) ((float) (srcX) * f), (double) ((float) (srcY) * f1)).endVertex();
        tessellator.draw();
    }

    public static void glTexturedRect(double x, double y, double z, double srcX, double srcY, double width, double height) {
        glTexturedRect(x, y, z, width, height, srcX, srcY, width, height);
    }

    public static void glTexturedRect(double x, double y, double width, double height, double srcX, double srcY, double srcWidth, double srcHeight) {
        glTexturedRect(x, y, 0, width, height, srcX, srcY, srcWidth, srcHeight);
    }

    public static void glTexturedRect(double x, double y, double srcX, double srcY, double width, double height) {
        glTexturedRect(x, y, 0, srcX, srcY, width, height);
    }

    public static void addVertex(double x, double y, double z) {
        Tessellator.getInstance().getBuffer().pos(x, y, z).endVertex();
    }

    public static void addVertex(double x, double y, double z, double srcX, double srcY) {
        Tessellator.getInstance().getBuffer().pos(x, y, z).tex(srcX, srcY).endVertex();
    }

    public static void addVertex(double x, double y, double z, double srcX, double srcY, float red, float green, float blue, float alpha) {
        Tessellator.getInstance().getBuffer().pos(x, y, z).tex(srcX, srcY).color(red, green, blue, alpha).endVertex();
    }

    public static void begin() {
        begin(GL11.GL_QUADS, DefaultVertexFormats.POSITION_TEX_COLOR);
    }

    public static void begin(int glMode, VertexFormat format) {
        Tessellator.getInstance().getBuffer().begin(glMode, format);
    }

    public static void draw() {
        Tessellator.getInstance().draw();
    }

    public static void glRect(int x, int y, int width, int height) {
        Tessellator tessellator = Tessellator.getInstance();
        tessellator.getBuffer().begin(GL11.GL_QUADS, DefaultVertexFormats.POSITION);
        tessellator.getBuffer().pos((double) (x), (double) (y + height), 0.0D);
        tessellator.getBuffer().pos((double) (x + width), (double) (y + height), 0.0D);
        tessellator.getBuffer().pos((double) (x + width), (double) (y), 0.0D);
        tessellator.getBuffer().pos((double) (x), (double) (y), 0.0D);
        tessellator.draw();
    }

    public static void glAlphaTest(boolean flag) {
        if (flag) GlStateManager.enableAlpha();
        else GlStateManager.disableAlpha();
    }

    public static void alphaFunc(int src, int dst) {
        GlStateManager.alphaFunc(src, dst);
    }

    public static void glBlend(boolean flag) {
        if (flag) GlStateManager.enableBlend();
        else GlStateManager.disableBlend();
    }

    public static void blendFunc(int src, int dst) {
        GlStateManager.blendFunc(src, dst);
    }

    public static void tryBlendFuncSeparate(int a, int b, int c, int d) {
        GlStateManager.tryBlendFuncSeparate(a, b, c, d);
    }

    public static void depthMask(boolean flag) {
        GlStateManager.depthMask(flag);
    }

    public static void glDepthTest(boolean flag) {
        if (flag) GlStateManager.enableDepth();
        else GlStateManager.disableDepth();
    }

    public static void glDepthFunc(int flag) {
        GL11.glDepthFunc(flag);
    }

    public static void glRescaleNormal(boolean flag) {
        if (flag) GlStateManager.enableRescaleNormal();
        else GlStateManager.disableRescaleNormal();
    }

    public static void glTexture2D(boolean flag) {
        if (flag) GL11.glEnable(GL11.GL_TEXTURE_2D);
        else GL11.glDisable(GL11.GL_TEXTURE_2D);
    }

    public static void glCullFace(boolean flag) {
        if (flag) GlStateManager.enableCull();
        else GlStateManager.disableCull();
    }

    public static void glTranslatef(float x, float y, float z) {
        GlStateManager.translate(x, y, z);
    }

    public static void glNormal3f(float x, float y, float z) {
        GL11.glNormal3f(x, y, z);
    }

    public static void glRotatef(float angle, float x, float y, float z) {
        GlStateManager.rotate(angle, x, y, z);
    }

    public static void glScalef(float x, float y, float z) {
        GlStateManager.scale(x, y, z);
    }

    public static void lighting(boolean flag) {
        if (flag) GlStateManager.enableLighting();
        else GlStateManager.disableLighting();
    }

    public static void glStartUI(Minecraft mc) {
        mc.mcProfiler.startSection(SAOCore.MODID + "[ '" + SAOCore.NAME + "' ]");
    }

    public static void glEndUI(Minecraft mc) {
        mc.mcProfiler.endSection();
    }

    public static void start() { GlStateManager.pushMatrix(); }

    public static void end() {
        GlStateManager.popMatrix();
    }


    /**
     * returns an AABB with corners x1, y1, z1 and x2, y2, z2
     */
    public static AxisAlignedBB fromBounds(double x1, double y1, double z1, double x2, double y2, double z2) {
        double d0 = Math.min(x1, x2);
        double d1 = Math.min(y1, y2);
        double d2 = Math.min(z1, z2);
        double d3 = Math.max(x1, x2);
        double d4 = Math.max(y1, y2);
        double d5 = Math.max(z1, z2);
        return new AxisAlignedBB(d0, d1, d2, d3, d4, d5);
    }
}

File: src/main/java/com/saomc/saoui/GLCore.kt
/*
 * Copyright (C) 2016-2019 Arnaud 'Bluexin' Solé
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.saomc.saoui

import com.mojang.blaze3d.matrix.MatrixStack
import com.mojang.blaze3d.platform.GlStateManager
import com.mojang.blaze3d.systems.RenderSystem
import com.saomc.saoui.SAOCore.mc
import com.saomc.saoui.themes.ThemeLoader
import com.saomc.saoui.util.ColorUtil
import com.teamwizardry.librarianlib.core.util.kotlin.builder
import com.teamwizardry.librarianlib.math.Vec2d
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.FontRenderer
import net.minecraft.client.renderer.BufferBuilder
import net.minecraft.client.renderer.Tessellator
import net.minecraft.client.renderer.WorldVertexBufferUploader
import net.minecraft.client.renderer.texture.TextureManager
import net.minecraft.client.renderer.vertex.DefaultVertexFormats
import net.minecraft.client.renderer.vertex.VertexFormat
import net.minecraft.util.ResourceLocation
import net.minecraft.util.math.AxisAlignedBB
import net.minecraft.util.math.vector.Matrix4f
import net.minecraft.util.math.vector.Vector2f
import net.minecraft.util.math.vector.Vector3d
import net.minecraft.util.text.TranslationTextComponent
import org.lwjgl.opengl.GL11
import kotlin.math.max
import kotlin.math.min

object GLCore {

    val glFont: FontRenderer get() = mc.fontRenderer
    private val glTextureManager get() = mc.textureManager
    val tessellator: Tessellator
        get() = Tessellator.getInstance()
    val bufferBuilder: BufferBuilder = tessellator.buffer

    @JvmOverloads
    fun color(red: Float, green: Float, blue: Float, alpha: Float = 1f) {
        RenderSystem.blendColor(red, green, blue, alpha)
    }

    fun color(color: ColorUtil) {
        color(color.rgba)
    }

    fun color(rgba: Int) {
        val red = (rgba shr 24 and 0xFF).toFloat() / 0xFF
        val green = (rgba shr 16 and 0xFF).toFloat() / 0xFF
        val blue = (rgba shr 8 and 0xFF).toFloat() / 0xFF
        val alpha = (rgba and 0xFF).toFloat() / 0xFF

        color(red, green, blue, alpha)
    }

    fun color(rgba: Int, lightLevel: Float) {
        val red = (rgba shr 24 and 0xFF).toFloat() / 0xFF
        val green = (rgba shr 16 and 0xFF).toFloat() / 0xFF
        val blue = (rgba shr 8 and 0xFF).toFloat() / 0xFF
        val alpha = (rgba and 0xFF).toFloat() / 0xFF
        val light = max(lightLevel, 0.15f)


        color(red * light, green * light, blue * light, alpha)
    }

    private fun glFontColor(rgba: Int): Int {
        val alpha = rgba and 0xFF
        val red = rgba shr 24 and 0xFF
        val blue = rgba shr 8 and 0xFF
        val green = rgba shr 16 and 0xFF

        return alpha shl 24 or (red shl 16) or (blue shl 8) or green
    }

    @JvmOverloads
    fun glString(font: FontRenderer, matrixStack: MatrixStack, string: String, x: Int, y: Int, argb: Int, shadow: Boolean = false, centered: Boolean = false) {
        if (shadow)
            font.drawTextWithShadow(matrixStack, TranslationTextComponent(string), x.toFloat(), y.toFloat() - if (centered) font.FONT_HEIGHT / 2f else 0f, glFontColor(argb))
        else
            font.drawString(matrixStack, string, x.toFloat(), y.toFloat() - if (centered) font.FONT_HEIGHT / 2f else 0f, glFontColor(argb))
    }

    @JvmOverloads
    fun glString(matrixStack: MatrixStack, string: String, x: Int, y: Int, argb: Int, shadow: Boolean = false, centered: Boolean = false) {
        glString(glFont, matrixStack, string, x, y, argb, shadow, centered)
    }

    @JvmOverloads
    fun glString(matrixStack: MatrixStack,string: String, pos: Vec2d, argb: Int, shadow: Boolean = false, centered: Boolean = false) {
        glString(matrixStack, string, pos.x.toInt(), pos.y.toInt(), argb, shadow, centered)
    }

    fun setFont(mc: Minecraft, custom: Boolean) {
        if (mc.textureManager == null) return
        val font = ResourceLocation(SAOCore.MODID, "textures/${ThemeLoader.currentTheme}/ascii.png")
        val fontLocation: ResourceLocation =
        if (custom && checkTexture(font)){
            font
        }
        else {
            ResourceLocation("textures/font/ascii.png")
        }
        /*
        mc.fontRenderer = FontRenderer(mc.gameSettings, fontLocation, mc.textureManager, false)
        if (mc.gameSettings.language != null) {
            mc.fontRenderer.unicodeFlag = mc.isUnicode
            mc.fontRenderer.bidiFlag = mc.languageManager.isCurrentLanguageBidirectional
        }
        (mc.resourceManager as IReloadableResourceManager).registerReloadListener(mc.fontRenderer)*/
    }

    @JvmOverloads
    fun glStringWidth(string: String, font: FontRenderer? = glFont): Int {
        return font?.getStringWidth(string) ?: 0
    }

    @JvmOverloads
    fun glStringHeight(font: FontRenderer? = glFont): Int {
        return font?.FONT_HEIGHT ?: 0
    }

    @JvmOverloads
    fun glBindTexture(location: ResourceLocation, textureManager: TextureManager = glTextureManager) {
        textureManager.bindTexture(location)
    }

    /**
     * Checks to make sure the texture is valid, returning false means the texture is invalid.
     */
    fun checkTexture(location: ResourceLocation): Boolean {
        return try {
            mc.resourceManager.getResource(location)
            true
        } catch (e: Exception) {
            false
        }
    }

    @JvmOverloads
    fun glTexturedRectV2(pos: Vector3d, size: Vector2f, srcPos: Vector2f = Vector2f.ZERO, srcSize: Vector2f = size, textureSize: Vector2f = Vector2f(256f, 256f)) {
        glTexturedRectV2(
                x = pos.x, y = pos.y, z = pos.z,
                width = size.x, height = size.y,
                srcX = srcPos.x, srcY = srcPos.y,
                srcWidth = srcSize.x, srcHeight = srcSize.y,
                textureW = textureSize.x, textureH = textureSize.y
        )
    }

    @JvmOverloads
    @Deprecated("Use version with MatrixStack")
    fun glTexturedRectV2(x: Double, y: Double, z: Double = 0.0, width: Float, height: Float, srcX: Float = 0f, srcY: Float = 0f, srcWidth: Float = width, srcHeight: Float = height, textureW: Float = 256f, textureH: Float = 256f) {
        val f = 1f / textureW
        val f1 = 1f / textureH
        begin(GL11.GL_QUADS, DefaultVertexFormats.POSITION_TEX)
        bufferBuilder.pos(x, y + height, z).tex(srcX * f, srcY + srcHeight * f1).endVertex()
        bufferBuilder.pos(x + width, y + height, z).tex(srcX + srcWidth * f, srcY + srcHeight * f1).endVertex()
        bufferBuilder.pos(x + width, y, z).tex(srcX + srcWidth * f, srcY * f1).endVertex()
        bufferBuilder.pos(x, y, z).tex(srcX * f, srcY * f1).endVertex()
        bufferBuilder.finishDrawing()
        RenderSystem.enableAlphaTest()
        WorldVertexBufferUploader.draw(bufferBuilder)
        draw()
    }

    @JvmOverloads
    fun glTexturedRectV2(matrixStack: MatrixStack, x: Float, y: Float, z: Float = 0f, width: Float, height: Float, srcX: Float = 0f, srcY: Float = 0f, srcWidth: Float = width, srcHeight: Float = height, textureW: Float = 256f, textureH: Float = 256f) {
        val f = 1f / textureW
        val f1 = 1f / textureH
        begin(GL11.GL_QUADS, DefaultVertexFormats.POSITION_TEX)
        bufferBuilder.pos(matrixStack.last.matrix, x, y + height, z).tex(srcX * f, srcY + srcHeight * f1).endVertex()
        bufferBuilder.pos(matrixStack.last.matrix, x + width, y + height, z).tex(srcX + srcWidth * f, srcY + srcHeight * f1).endVertex()
        bufferBuilder.pos(matrixStack.last.matrix, x + width, y, z).tex(srcX + srcWidth * f, srcY * f1).endVertex()
        bufferBuilder.pos(matrixStack.last.matrix, x, y, z).tex(srcX * f, srcY * f1).endVertex()
        draw()
    }

    fun addVertex(x: Double, y: Double, z: Double) {
        bufferBuilder.pos(x, y, z).endVertex()
    }

    fun addVertex(x: Double, y: Double, z: Double, srcX: Float, srcY: Float) {
        bufferBuilder.pos(x, y, z).tex(srcX, srcY).endVertex()
    }

    fun addVertex(x: Double, y: Double, z: Double, srcX: Float, srcY: Float, red: Float, green: Float, blue: Float, alpha: Float) {
        bufferBuilder.pos(x, y, z).tex(srcX, srcY).color(red, green, blue, alpha).endVertex()
    }

    fun addVertex(x: Double, y: Double, z: Double, red: Float, green: Float, blue: Float, alpha: Float) {
        bufferBuilder.pos(x, y, z).color(red, green, blue, alpha).endVertex()
    }

    fun addVertex(x: Float, y: Float, z: Float, srcX: Float, srcY: Float, red: Float, green: Float, blue: Float, alpha: Float) {
        bufferBuilder.normal(x, y, z).tex(srcX, srcY).color(red, green, blue, alpha).endVertex()
    }

    fun addVertex(matrix: Matrix4f, x: Float, y: Float, z: Float) {
        bufferBuilder.pos(matrix, x, y, z).endVertex()
    }

    fun addVertex(matrix: Matrix4f, x: Float, y: Float, z: Float, srcX: Float, srcY: Float) {
        bufferBuilder.pos(matrix, x, y, z).tex(srcX, srcY).endVertex()
    }

    fun addVertex(matrix: Matrix4f, x: Float, y: Float, z: Float, srcX: Float, srcY: Float, red: Float, green: Float, blue: Float, alpha: Float) {
        bufferBuilder.pos(matrix, x, y, z).tex(srcX, srcY).color(red, green, blue, alpha).endVertex()
    }

    fun addVertex(matrix: Matrix4f, x: Float, y: Float, z: Float, red: Float, green: Float, blue: Float, alpha: Float) {
        bufferBuilder.pos(matrix, x, y, z).color(red, green, blue, alpha).endVertex()
    }

    @JvmOverloads
    fun begin(glMode: Int = GL11.GL_QUADS, format: VertexFormat = DefaultVertexFormats.POSITION_TEX_COLOR) {
        bufferBuilder.begin(glMode, format)
    }

    fun draw() {
        tessellator.draw()
    }

    @Deprecated("Use below")
    fun glRect(x: Int, y: Int, width: Int, height: Int) {
        begin(GL11.GL_QUADS, DefaultVertexFormats.POSITION)
        bufferBuilder.pos(x.toDouble(), (y + height).toDouble(), 0.0)
        bufferBuilder.pos((x + width).toDouble(), (y + height).toDouble(), 0.0)
        bufferBuilder.pos((x + width).toDouble(), y.toDouble(), 0.0)
        bufferBuilder.pos(x.toDouble(), y.toDouble(), 0.0)
        draw()
    }

    fun glRect(matrix: Matrix4f, x: Int, y: Int, width: Int, height: Int) {
        begin(GL11.GL_QUADS, DefaultVertexFormats.POSITION)
        bufferBuilder.pos(x.toDouble(), (y + height).toDouble(), 0.0)
        bufferBuilder.pos((x + width).toDouble(), (y + height).toDouble(), 0.0)
        bufferBuilder.pos((x + width).toDouble(), y.toDouble(), 0.0)
        bufferBuilder.pos(x.toDouble(), y.toDouble(), 0.0)
        draw()
    }



    private fun innerBlit(
        matrix: Matrix4f,
        x1: Int,
        x2: Int,
        y1: Int,
        y2: Int,
        blitOffset: Int,
        minU: Float,
        maxU: Float,
        minV: Float,
        maxV: Float
    ) {
        val bufferbuilder = Tessellator.getInstance().buffer
        bufferbuilder.begin(7, DefaultVertexFormats.POSITION_TEX)
        bufferbuilder.pos(matrix, x1.toFloat(), y2.toFloat(), blitOffset.toFloat()).tex(minU, maxV).endVertex()
        bufferbuilder.pos(matrix, x2.toFloat(), y2.toFloat(), blitOffset.toFloat()).tex(maxU, maxV).endVertex()
        bufferbuilder.pos(matrix, x2.toFloat(), y1.toFloat(), blitOffset.toFloat()).tex(maxU, minV).endVertex()
        bufferbuilder.pos(matrix, x1.toFloat(), y1.toFloat(), blitOffset.toFloat()).tex(minU, minV).endVertex()
        bufferbuilder.finishDrawing()
        RenderSystem.enableAlphaTest()
        WorldVertexBufferUploader.draw(bufferbuilder)
    }

    /**
     * Draws a rectangle with a vertical gradient between the specified colors (ARGB format). Args : x1, y1, x2, y2,
     * topColor, bottomColor
     */
    fun drawGradientRect(left: Double, top: Double, right: Double, bottom: Double, zLevel: Double, startColor: Int, endColor: Int) {
        val f = (startColor shr 24 and 255).toFloat() / 255.0f
        val f1 = (startColor shr 16 and 255).toFloat() / 255.0f
        val f2 = (startColor shr 8 and 255).toFloat() / 255.0f
        val f3 = (startColor and 255).toFloat() / 255.0f
        val f4 = (endColor shr 24 and 255).toFloat() / 255.0f
        val f5 = (endColor shr 16 and 255).toFloat() / 255.0f
        val f6 = (endColor shr 8 and 255).toFloat() / 255.0f
        val f7 = (endColor and 255).toFloat() / 255.0f
        RenderSystem.disableTexture()
        RenderSystem.enableBlend()
        RenderSystem.disableAlphaTest()
        RenderSystem.blendFuncSeparate(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA, GlStateManager.SourceFactor.ONE, GlStateManager.DestFactor.ZERO)
        RenderSystem.shadeModel(7425)
        begin(7, DefaultVertexFormats.POSITION_COLOR)
        bufferBuilder.pos(right, top, zLevel).color(f1, f2, f3, f).endVertex()
        bufferBuilder.pos(left, top, zLevel).color(f1, f2, f3, f).endVertex()
        bufferBuilder.pos(left, bottom, zLevel).color(f5, f6, f7, f4).endVertex()
        bufferBuilder.pos(right, bottom, zLevel).color(f5, f6, f7, f4).endVertex()
        draw()
        RenderSystem.shadeModel(7424)
        RenderSystem.disableBlend()
        RenderSystem.enableAlphaTest()
        RenderSystem.enableTexture()
    }

    /**
     * Draws a rectangle with a vertical gradient between the specified colors (ARGB format). Args : x1, y1, x2, y2,
     * topColor, bottomColor
     */
    fun drawGradientRect(mstack: Matrix4f, left: Float, top: Float, right: Float, bottom: Float, zLevel: Float, startColor: Int, endColor: Int = startColor) {
        val f = (startColor shr 24 and 255).toFloat() / 255.0f
        val f1 = (startColor shr 16 and 255).toFloat() / 255.0f
        val f2 = (startColor shr 8 and 255).toFloat() / 255.0f
        val f3 = (startColor and 255).toFloat() / 255.0f
        val f4 = (endColor shr 24 and 255).toFloat() / 255.0f
        val f5 = (endColor shr 16 and 255).toFloat() / 255.0f
        val f6 = (endColor shr 8 and 255).toFloat() / 255.0f
        val f7 = (endColor and 255).toFloat() / 255.0f
        RenderSystem.disableTexture()
        RenderSystem.enableBlend()
        RenderSystem.disableAlphaTest()
        RenderSystem.blendFuncSeparate(GlStateManager.SourceFactor.SRC_ALPHA, GlStateManager.DestFactor.ONE_MINUS_SRC_ALPHA, GlStateManager.SourceFactor.ONE, GlStateManager.DestFactor.ZERO)
        RenderSystem.shadeModel(7425)
        begin(7, DefaultVertexFormats.POSITION_COLOR)
        bufferBuilder.pos(mstack, right, top, zLevel).color(f1, f2, f3, f).endVertex()
        bufferBuilder.pos(mstack, left, top, zLevel).color(f1, f2, f3, f).endVertex()
        bufferBuilder.pos(mstack, left, bottom, zLevel).color(f5, f6, f7, f4).endVertex()
        bufferBuilder.pos(mstack, right, bottom, zLevel).color(f5, f6, f7, f4).endVertex()
        draw()
        RenderSystem.shadeModel(7424)
        RenderSystem.disableBlend()
        RenderSystem.enableAlphaTest()
        RenderSystem.enableTexture()
    }


    @Deprecated("Use RenderType",
        ReplaceWith("RenderType", "net.minecraft.client.renderer")
    )
    fun glAlphaTest(flag: Boolean) {
        if (flag)
            RenderSystem.enableAlphaTest()
        else
            RenderSystem.disableAlphaTest()
    }

    @Deprecated("Use RenderType",
        ReplaceWith("RenderType", "net.minecraft.client.renderer")
    )
    fun alphaFunc(src: Int, dst: Int) {
        RenderSystem.alphaFunc(src, dst.toFloat())
    }

    fun glBlend(flag: Boolean) {
        if (flag)
            RenderSystem.enableBlend()
        else
            RenderSystem.disableBlend()
    }

    fun blendFunc(src: Int, dst: Int) {
        RenderSystem.blendFunc(src, dst)
    }

    fun tryBlendFuncSeparate(a: Int, b: Int, c: Int, d: Int) {
        RenderSystem.blendFuncSeparate(a, b, c, d)
    }

    fun depthMask(flag: Boolean) {
        RenderSystem.depthMask(flag)
    }

    fun depth(flag: Boolean) {
        if (flag)
            RenderSystem.enableDepthTest()
        else
            RenderSystem.disableDepthTest()
    }

    fun glDepthFunc(flag: Int) {
        RenderSystem.depthFunc(flag)
    }

    @Deprecated("Use RenderType",
        ReplaceWith("RenderType", "net.minecraft.client.renderer")
    )
    fun glRescaleNormal(flag: Boolean) {
        if (flag)
            RenderSystem.enableRescaleNormal()
        else
            RenderSystem.disableRescaleNormal()
    }

    fun glTexture2D(flag: Boolean) {
        if (flag)
            RenderSystem.enableTexture()
        else
            RenderSystem.disableTexture()
    }

    fun glCullFace(flag: Boolean) {
        if (flag)
            RenderSystem.enableCull()
        else
            RenderSystem.disableCull()
    }

    @Deprecated("Use MatrixStack",
        ReplaceWith("MatrixStack.translate(x, y, z)", "com.mojang.blaze3d.matrix.MatrixStack")
    )
    fun glTranslatef(x: Float, y: Float, z: Float) {
        RenderSystem.translatef(x, y, z)
    }

    fun glNormal3f(x: Float, y: Float, z: Float) {
        RenderSystem.normal3f(x, y, z)
    }

    @Deprecated("Use MatrixStack",
        ReplaceWith("MatrixStack.rotate(quaternion)", "com.mojang.blaze3d.matrix.MatrixStack")
    )
    fun glRotatef(angle: Float, x: Float, y: Float, z: Float) {
        RenderSystem.rotatef(angle, x, y, z)
    }

    @Deprecated("Use MatrixStack",
        ReplaceWith("MatrixStack.scale(x, y, z)", "com.mojang.blaze3d.matrix.MatrixStack")
    )
    fun glScalef(x: Float, y: Float, z: Float) {
        RenderSystem.scalef(x, y, z)
    }

    @Deprecated("Use RenderType",
        ReplaceWith("RenderType", "net.minecraft.client.renderer")
    )
    fun lighting(flag: Boolean) {
        if (flag)
            RenderSystem.enableLighting()
        else
            RenderSystem.disableLighting()
    }

    @Deprecated("Use MatrixStack",
        ReplaceWith("MatrixStack.push()", "com.mojang.blaze3d.matrix.MatrixStack")
    )
    fun pushMatrix() {
        RenderSystem.pushMatrix()
    }

    @Deprecated("Use MatrixStack",
        ReplaceWith("MatrixStack.pop()", "com.mojang.blaze3d.matrix.MatrixStack")
    )
    fun popMatrix() {
        RenderSystem.popMatrix()
    }


    /**
     * returns an AABB with corners x1, y1, z1 and x2, y2, z2
     */
    fun fromBounds(x1: Double, y1: Double, z1: Double, x2: Double, y2: Double, z2: Double): AxisAlignedBB {
        val d0 = min(x1, x2)
        val d1 = min(y1, y2)
        val d2 = min(z1, z2)
        val d3 = max(x1, x2)
        val d4 = max(y1, y2)
        val d5 = max(z1, z2)
        return AxisAlignedBB(d0, d1, d2, d3, d4, d5)
    }

    @Deprecated("Use MatrixStack",
        ReplaceWith("MatrixStack.translate(x, y, z)", "com.mojang.blaze3d.matrix.MatrixStack")
    )
    fun translate(x: Double, y: Double, z: Double) {
        RenderSystem.translated(x, y, z)
    }

    @Deprecated("Use MatrixStack",
        ReplaceWith("MatrixStack.translate(x, y, z)", "com.mojang.blaze3d.matrix.MatrixStack")
    )
    fun translate(x: Float, y: Float, z: Float) {
        RenderSystem.translatef(x, y, z)
    }

    @Deprecated("Use MatrixStack",
        ReplaceWith("MatrixStack.scale(x, y, z)", "com.mojang.blaze3d.matrix.MatrixStack")
    )
    fun scale(x: Double, y: Double, z: Double) {
        RenderSystem.scaled(x, y, z)
    }

    @Deprecated("Use MatrixStack",
        ReplaceWith("MatrixStack.scale(x, y, z)", "com.mojang.blaze3d.matrix.MatrixStack")
    )
    fun scale(x: Float, y: Float, z: Float) {
        RenderSystem.scalef(x, y, z)
    }
}
